File: android-agent/app/src/main/kotlin/com/remoteagent/cryptokit/ReplayGuard.kt
package com.remoteagent.cryptokit

/**
 * Monotonic per-direction sequence replay protection.
 */
class ReplayGuard {
    private var lastSeq: Long = 0
    private val seen = HashSet<Long>()

    @Synchronized
    fun accept(seq: Long) {
        require(seq > 0) { "sequence must be > 0" }
        require(seq !in seen) { "duplicate sequence" }
        require(seq > lastSeq) { "replayed or older sequence" }
        seen.add(seq)
        lastSeq = seq
        if (seen.size > 4096) {
            seen.clear()
            seen.add(seq)
        }
    }

    @Synchronized
    fun reset() {
        lastSeq = 0
        seen.clear()
    }
}

File: server/internal/websocket/replay_guard.go
package websocket

import "errors"

// ReplayGuard enforces monotonic per-direction sequence numbers.
type ReplayGuard struct {
	last uint64
	seen map[uint64]struct{}
}

func NewReplayGuard() *ReplayGuard {
	return &ReplayGuard{seen: make(map[uint64]struct{})}
}

func (g *ReplayGuard) Accept(seq uint64) error {
	if seq == 0 {
		return errors.New("sequence must be > 0")
	}
	if _, ok := g.seen[seq]; ok {
		return errors.New("duplicate sequence")
	}
	if seq <= g.last {
		return errors.New("replayed or older sequence")
	}
	g.seen[seq] = struct{}{}
	g.last = seq
	if len(g.seen) > 4096 {
		g.seen = map[uint64]struct{}{seq: {}}
	}
	return nil
}

func (g *ReplayGuard) Reset() {
	g.last = 0
	g.seen = make(map[uint64]struct{})
}
